File: app/src/main/java/com/afollestad/photoaffix/adapters/PhotoGridAdapter.java
package com.afollestad.photoaffix.adapters;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.afollestad.dragselectrecyclerview.DragSelectRecyclerViewAdapter;
import com.afollestad.photoaffix.R;
import com.afollestad.photoaffix.data.Photo;
import com.afollestad.photoaffix.data.PhotoHolder;
import com.afollestad.photoaffix.ui.MainActivity;
import com.bumptech.glide.Glide;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

/** @author Aidan Follestad (afollestad) */
public class PhotoGridAdapter
    extends DragSelectRecyclerViewAdapter<PhotoGridAdapter.PhotoViewHolder> {

  private final MainActivity context;
  private Photo[] photos;

  public PhotoGridAdapter(MainActivity context) {
    this.context = context;
  }

  @Override
  public void saveInstanceState(Bundle out) {
    super.saveInstanceState(out);
    if (photos != null) {
      out.putSerializable("photos", new PhotoHolder(photos));
    }
  }

  @Override
  public void restoreInstanceState(Bundle in) {
    super.restoreInstanceState(in);
    if (in != null && in.containsKey("photos")) {
      PhotoHolder ph = (PhotoHolder) in.getSerializable("photos");
      if (ph != null) {
        setPhotos(ph.photos);
      }
    }
  }

  public void shiftSelections() {
    for (int i = 1; i < getItemCount() - 1; i++) {
      boolean currentSelected = isIndexSelected(i);
      if (currentSelected) {
        setSelected(i + 1, true);
        setSelected(i, false);
        i++;
      }
    }
  }

  public void setPhotos(Photo[] photos) {
    this.photos = photos;
    notifyDataSetChanged();
  }

  public Photo[] getSelectedPhotos() {
    Integer[] indices = getSelectedIndices();
    ArrayList<Photo> selected = new ArrayList<>();
    for (Integer index : indices) {
      if (index < 0) {
        continue;
      }
      selected.add(photos[index - 1]);
    }
    return selected.toArray(new Photo[selected.size()]);
  }

  @Override
  public int getItemViewType(int position) {
    return position == 0 ? 0 : 1;
  }

  @Override
  public PhotoViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    View v =
        LayoutInflater.from(parent.getContext())
            .inflate(
                viewType == 0 ? R.layout.griditem_browse : R.layout.griditem_photo, parent, false);
    return new PhotoViewHolder(v);
  }

  @SuppressWarnings("ConstantConditions")
  @Override
  public void onBindViewHolder(PhotoViewHolder holder, int position) {
    super.onBindViewHolder(holder, position);
    if (context == null || context.isFinishing()) {
      return;
    }
    if (position == 0) {
      return;
    }

    Glide.with(context).load(photos[position - 1].getUri()).into(holder.image);

    if (isIndexSelected(position)) {
      holder.check.setVisibility(View.VISIBLE);
      holder.circle.setActivated(true);
      holder.image.setActivated(true);
    } else {
      holder.check.setVisibility(View.GONE);
      holder.circle.setActivated(false);
      holder.image.setActivated(false);
    }
  }

  @Override
  protected boolean isIndexSelectable(int index) {
    return index > 0;
  }

  @Override
  public int getItemCount() {
    return photos != null ? photos.length + 1 : 0;
  }

  class PhotoViewHolder extends RecyclerView.ViewHolder {

    @Nullable
    @BindView(R.id.image)
    ImageView image;

    @Nullable
    @BindView(R.id.check)
    View check;

    @Nullable
    @BindView(R.id.circle)
    View circle;

    PhotoViewHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);

      itemView.setOnClickListener(
          v -> {
            if (getAdapterPosition() == 0) {
              context.browseExternalPhotos();
              return;
            }
            toggleSelected(getAdapterPosition());
          });

      if (image != null) {
        itemView.setOnLongClickListener(
            v -> {
              toggleSelected(getAdapterPosition());
              context.list.setDragSelectActive(true, getAdapterPosition());
              return false;
            });
      }
    }
  }
}

File: app/src/main/java/com/afollestad/photoaffix/adapters/PhotoGridAdapter.kt
/*
 * Licensed under Apache-2.0
 *
 * Designed and developed by Aidan Follestad (@afollestad)
 */
package com.afollestad.photoaffix.adapters

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.afollestad.dragselectrecyclerview.DragSelectReceiver
import com.afollestad.dragselectrecyclerview.DragSelectTouchListener
import com.afollestad.photoaffix.R
import com.afollestad.photoaffix.views.MainActivity
import com.afollestad.photoaffix.data.Photo
import com.afollestad.photoaffix.data.PhotoHolder
import com.afollestad.photoaffix.utilities.loadImage
import kotlinx.android.synthetic.main.griditem_photo.view.check
import kotlinx.android.synthetic.main.griditem_photo.view.circle
import kotlinx.android.synthetic.main.griditem_photo.view.image

typealias SelectionListener = (position: Int, count: Int) -> Unit

class PhotoViewHolder(
  itemView: View,
  private val adapter: PhotoGridAdapter
) : RecyclerView.ViewHolder(itemView) {

  init {
    val context = itemView.context as MainActivity
    itemView.setOnClickListener {
      if (adapterPosition == 0) {
        context.browseExternalPhotos()
      } else {
        adapter.toggleSelected(adapterPosition)
      }
    }

    if (itemView.image != null) {
      itemView.setOnLongClickListener { v ->
        adapter.toggleSelected(adapterPosition)
        adapter.dragListener?.setIsActive(true, adapterPosition)
        false
      }
    }
  }

  fun bind(photo: Photo) {
    itemView.image.loadImage(photo.uri)

    if (adapter.isSelected(adapterPosition)) {
      itemView.check.visibility = VISIBLE
      itemView.circle.isActivated = true
      itemView.image.isActivated = true
    } else {
      itemView.check.visibility = GONE
      itemView.circle.isActivated = false
      itemView.image.isActivated = false
    }
  }
}

/** @author Aidan Follestad (afollestad) */
class PhotoGridAdapter(val context: MainActivity) : RecyclerView.Adapter<PhotoViewHolder>(),
    DragSelectReceiver {

  companion object {
    private const val KEY_PHOTOS = "photos"
    private const val KEY_SELECTED_INDICES = "selectedIndices"
  }

  val selectedPhotos: List<Photo>
    get() {
      return selectedIndices.map { photos[it - 1] }
    }
  var dragListener: DragSelectTouchListener? = null

  private var photos = listOf<Photo>()
  private var selectedIndices = mutableListOf<Int>()
  private var selectionListener: SelectionListener? = null

  fun onSelection(selection: SelectionListener) {
    this.selectionListener = selection
  }

  fun hasSelection() = selectedIndices.isNotEmpty()

  fun toggleSelected(index: Int) = setSelected(index, !isSelected(index))

  fun shiftSelections() {
    var i = 1
    while (i < itemCount - 1) {
      val currentSelected = isSelected(i)
      if (currentSelected) {
        setSelected(i + 1, true)
        setSelected(i, false)
        i++
      }
      i++
    }
  }

  fun clearSelected() {
    selectedIndices.clear()
    this.selectionListener?.invoke(0, 0)
    notifyDataSetChanged()
  }

  fun setPhotos(photos: List<Photo>) {
    this.photos = photos
    notifyDataSetChanged()
  }

  fun saveInstanceState(out: Bundle) {
    if (photos.isNotEmpty()) {
      out.putSerializable(KEY_PHOTOS, PhotoHolder(photos))
      out.putIntArray(KEY_SELECTED_INDICES, selectedIndices.toIntArray())
    }
  }

  fun restoreInstanceState(savedState: Bundle?) {
    if (savedState?.containsKey(KEY_PHOTOS) == true) {
      val ph = savedState.getSerializable(KEY_PHOTOS) as? PhotoHolder ?: return
      setPhotos(ph.photos)
    }
    savedState?.getIntArray(KEY_SELECTED_INDICES)
        ?.let {
          this.selectedIndices = it.toMutableList()
          notifyDataSetChanged()
        }
  }

  override fun isSelected(index: Int) = selectedIndices.contains(index)

  override fun isIndexSelectable(index: Int) = index > 0

  override fun setSelected(
    index: Int,
    selected: Boolean
  ) {
    if (selected && !isSelected(index) && isIndexSelectable(index)) {
      selectedIndices.add(index)
    } else if (!selected) {
      selectedIndices.remove(index)
    }
    this.selectionListener?.invoke(index, selectedIndices.size)
    notifyItemChanged(index)
  }

  override fun getItemViewType(position: Int): Int {
    return if (position == 0) 0 else 1
  }

  override fun onCreateViewHolder(
    parent: ViewGroup,
    viewType: Int
  ): PhotoViewHolder {
    val res =
      if (viewType == 0) R.layout.griditem_browse
      else R.layout.griditem_photo
    val view = LayoutInflater.from(parent.context)
        .inflate(res, parent, false)
    return PhotoViewHolder(view, this)
  }

  override fun onBindViewHolder(
    holder: PhotoViewHolder,
    position: Int
  ) {
    if (context.isFinishing || position == 0) return
    val photo = photos[position - 1]
    holder.bind(photo)
  }

  override fun getItemCount() = photos.size + 1
}
